domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentProgressController.kt

```diff
@@ -157,13 +171,31 @@ class QuestionAssessmentProgressController @Inject constructor(
             if (!progress.isAssessmentCompleted()) {
               // Only push the next state if the assessment isn't completed.
               progress.stateDeck.pushState(progress.getNextState(), prohibitSameStateName = false)
+              // Reset the hintState if pending top state has changed.
+              progress.hintState = hintHandler.reset()
             } else {
               // Otherwise, push a synthetic state for the end of the session.
               progress.stateDeck.pushState(
                 State.getDefaultInstance(),
                 prohibitSameStateName = false
               )
             }
+          } else {
+            // Schedule a new hints or solution or show a new hint or solution immediately based on
```

Comment:
Do we need tests for this scenario in QuestionAssessmentProgressControllerTest?

Reply:
We have another issue tracking this (#1273). The state fragment also uses the same logic for hints and solutions and we do have tests in state fragment local tests that test this behaviour.